// rollup-plugin-preserve-shebang/dist/index.js
'use strict';

exports.__esModule = true;
exports.default = shebangPlugin;

var _magicString = require('magic-string');

var _magicString2 = _interopRequireDefault(_magicString);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function shebangPlugin() {
	var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
	    shebang = _ref.shebang,
	    entry = _ref.entry;

	function processEntry(entry) {
		if (entry) {
			var contents = _fs2.default.readFileSync(entry, 'utf8');
			var matches = contents.match(/^\s*(#!.*)/);
			shebang = matches && matches[1] || false;
		}
	}
	processEntry(entry);

	return {
		name: 'preserve-shebang',
		options: function options(_options) {
			if (!entry) {
				entry = _options.input;
				processEntry(entry);
			}
			return _options;
		},
		transformBundle: function transformBundle(code, _ref2) {
			var format = _ref2.format,
			    sourcemap = _ref2.sourcemap;

			if (!shebang) return;

			var str = new _magicString2.default(code);
			str.prepend(shebang + '\n');
			return {
				code: str.toString(),
				map: sourcemap ? str.generateMap({ hires: true }) : undefined
			};
		}
	};
}
